File: parse-rdb.js
/**
 * Minimal Redis RDB 6.0 parser
 * Extracts all keys related to a given username
 * Usage: node parse-rdb.js <rdb-file> [username]
 */

const fs = require('fs');
const path = require('path');

const RDB_OPCODE_EOF = 0xFF;
const RDB_OPCODE_SELECTDB = 0xFE;
const RDB_OPCODE_EXPIRETIME_MS = 0xFC;
const RDB_OPCODE_EXPIRETIME = 0xFD;
const RDB_TYPE_STRING = 0;
const RDB_TYPE_LIST = 1;
const RDB_TYPE_SET = 2;
const RDB_TYPE_ZSET = 3;
const RDB_TYPE_HASH = 4;
const RDB_TYPE_ZSET_2 = 5;

class RDBParser {
  constructor(buffer) {
    this.buf = buffer;
    this.pos = 0;
    this.db = 0;
    this.keys = {};
  }

  readByte() {
    if (this.pos >= this.buf.length) throw new Error('Unexpected EOF at ' + this.pos);
    return this.buf[this.pos++];
  }

  readBytes(n) {
    if (this.pos + n > this.buf.length) throw new Error('Unexpected EOF');
    const slice = this.buf.slice(this.pos, this.pos + n);
    this.pos += n;
    return slice;
  }

  // Returns { value: number|bigint, isEncoded: boolean }
  readLength() {
    const first = this.readByte();
    const encType = (first & 0xC0) >> 6;

    if (encType === 0) {
      return { value: first & 0x3F, isEncoded: false };
    } else if (encType === 1) {
      const next = this.readByte();
      return { value: ((first & 0x3F) << 8) | next, isEncoded: false };
    } else if (encType === 3) {
      const bytes = this.readBytes(8);
      let value = 0n;
      for (let i = 7; i >= 0; i--) {
        value = (value << 8n) | BigInt(bytes[i]);
      }
      return { value: Number(value), isEncoded: false };
    } else {
      // encType === 2 : special encoding in remaining 6 bits
      const enc = first & 0x3F;
      if (enc === 0) {
        const bytes = this.readBytes(4);
        return { value: new DataView(bytes.buffer).getInt32(0, true), isEncoded: true };
      } else if (enc === 1) {
        const bytes = this.readBytes(8);
        return { value: new DataView(bytes.buffer).getBigInt64(0, true), isEncoded: true };
      } else if (enc === 2) {
        const { value: len } = this.readLength();
        const str = this.readBytes(len).toString('utf8');
        return { value: parseInt(str, 10), isEncoded: true };
      }
    }
    throw new Error('Invalid length encoding');
  }

  readString() {
    const { value, isEncoded } = this.readLength();
    if (isEncoded) return String(value);
    return this.readBytes(value).toString('utf8');
  }

  readDouble() {
    const bytes = this.readBytes(8);
    return new DataView(bytes.buffer).getFloat64(0, true);
  }

  parseIntset() {
    const data = this.readBytes(this.readLength().value);
    const encoding = data.readUInt32LE(0);
    const count = data.readUInt32LE(4);
    const result = [];
    for (let i = 0; i < count; i++) {
      if (encoding === 2) result.push(data.readInt32LE(8 + i * 4));
      else if (encoding === 1) result.push(data.readInt16LE(8 + i * 2));
      else result.push(data.readInt8(8 + i));
    }
    return result;
  }

  parseHash() {
    const { value: count } = this.readLength();
    const result = {};
    for (let i = 0; i < count; i++) {
      const field = this.readString();
      result[field] = this.readString();
    }
    return result;
  }

  parseSet() {
    const { value: count } = this.readLength();
    const result = [];
    for (let i = 0; i < count; i++) result.push(this.readString());
    return result;
  }

  parseZset() {
    const { value: count } = this.readLength();
    const result = [];
    for (let i = 0; i < count; i++) {
      const member = this.readString();
      const score = this.readDouble();
      result.push({ member, score });
    }
    return result;
  }

  parseList() {
    const { value: count } = this.readLength();
    const result = [];
    for (let i = 0; i < count; i++) result.push(this.readString());
    return result;
  }

  parse() {
    // Magic: REDIS
    const magic = this.readBytes(5).toString('ascii');
    if (magic !== 'REDIS') throw new Error('Invalid RDB magic: ' + magic);

    const verStr = this.readBytes(4).toString('ascii');
    const version = parseInt(verStr, 10);
    console.log(`RDB version: ${version}`);

    while (this.pos < this.buf.length) {
      const opcode = this.readByte();

      if (opcode === RDB_OPCODE_EOF) {
        console.log('Reached EOF marker');
        break;
      }

      if (opcode === RDB_OPCODE_SELECTDB) {
        const { value: dbNum } = this.readLength();
        this.db = dbNum;
        continue;
      }

      let expire = null;
      if (opcode === RDB_OPCODE_EXPIRETIME_MS) {
        expire = this.readBytes(8); // 8-byte MS timestamp
        opcode = this.readByte();
      } else if (opcode === RDB_OPCODE_EXPIRETIME) {
        expire = this.readBytes(4); // 4-byte UNIX timestamp
        opcode = this.readByte();
      }

      this.parseKeyValue(opcode, expire);
    }

    return this.keys;
  }

  parseKeyValue(type, expire) {
    const key = this.readString();
    let value, valueType;

    switch (type) {
      case RDB_TYPE_STRING:
        value = this.readString();
        valueType = 'string';
        break;
      case RDB_TYPE_HASH:
        value = this.parseHash();
        valueType = 'hash';
        break;
      case RDB_TYPE_SET:
        value = this.parseSet();
        valueType = 'set';
        break;
      case RDB_TYPE_ZSET:
      case RDB_TYPE_ZSET_2:
        value = this.parseZset();
        valueType = 'zset';
        break;
      case RDB_TYPE_LIST:
        value = this.parseList();
        valueType = 'list';
        break;
      default:
        console.warn(`  [db${this.db}] Skipping unknown type ${type} for key: ${key}`);
        return;
    }

    const fullKey = `db${this.db}:${key}`;
    this.keys[fullKey] = { key, db: this.db, type: valueType, value, expire };
  }
}

// ---- Main ----
const rdbPath = process.argv[2] || 'E:/Users/Downloads/118ff503-a34e-4401-9d01-1e9e42221165.rdb';
const targetUsers = process.argv[3] ? process.argv[3].split(',') : ['7788'];

console.log(`Parsing RDB file: ${rdbPath}`);
console.log(`Looking for users: ${targetUsers.join(', ')}\n`);

const buffer = fs.readFileSync(rdbPath);
const parser = new RDBParser(buffer);

try {
  const allKeys = parser.parse();
  console.log(`\nTotal keys found: ${Object.keys(allKeys).length}`);
  console.log(`\n=== Keys related to specified users: ===\n`);

  const userKeys = {};
  for (const [fullKey, data] of Object.entries(allKeys)) {
    const isRelevant = targetUsers.some(u => data.key.includes(u)) || data.key.includes('admin') || data.key.includes('config');
    if (isRelevant) {
      userKeys[fullKey] = data;
      console.log(`[${data.type}] ${data.key}`);
      if (data.type === 'string') {
        const val = String(data.value);
        console.log(`  = ${val.length > 300 ? val.substring(0, 300) + '...' : val}`);
      } else {
        console.log(`  = ${JSON.stringify(data.value).substring(0, 500)}`);
      }
      console.log('');
    }
  }

  if (Object.keys(userKeys).length === 0) {
    console.log('No matching keys found.');
    console.log('\nSample keys (first 30):');
    for (const [fullKey, data] of Object.entries(allKeys).slice(0, 30)) {
      console.log(`  [${data.type}] ${data.key}`);
    }
  }

  // Save extracted keys
  const outputBase = path.join(path.dirname(rdbPath), 'rdb-extract');
  const outputPath = `${outputBase}.json`;
  fs.writeFileSync(outputPath, JSON.stringify(userKeys, null, 2));
  console.log(`\nSaved to: ${outputPath}`);

  // Also save all keys summary
  const summary = {};
  for (const [fullKey, data] of Object.entries(allKeys)) {
    summary[data.key] = data.type;
  }
  const summaryPath = path.join(path.dirname(rdbPath), 'rdb-all-keys.json');
  fs.writeFileSync(summaryPath, JSON.stringify(summary, null, 2));
  console.log(`Key summary saved to: ${summaryPath}`);

} catch (err) {
  console.error('Parse error:', err.message);
  console.error('At position:', parser.pos);
  const start = Math.max(0, parser.pos - 20);
  const end = Math.min(buffer.length, parser.pos + 20);
  console.error('Nearby bytes:', buffer.slice(start, end).toString('hex'));
  process.exit(1);
}

File: parse_rdb.py
#!/usr/bin/env python3
"""Redis RDB 12 (Redis 7.x) parser - handles AUX fields and all types."""
import sys, json, struct

RDB_EOF           = 0xFF
RDB_SELECTDB       = 0xFE
RDB_EXPIRETIME_MS  = 0xFC
RDB_EXPIRETIME     = 0xFD
RDB_RESIZEDB       = 0xFB
RDB_AUX            = 0xFA

T_STRING    = 0
T_LIST      = 1
T_SET       = 2
T_ZSET      = 3
T_HASH      = 4
T_ZSET_2    = 5
T_MODULE    = 6
T_MODULE_2  = 7
T_HASH_ZIPMAP   = 8
T_LIST_ZIPLIST  = 9
T_SET_INTSET    = 10
T_ZSET_ZIPLIST  = 11
T_HASH_ZIPLIST  = 12
T_LIST_QUICKLIST = 13
T_STREAM_LISTPACKS = 14
T_HASH_LISTPACK    = 15
T_ZSET_LISTPACK    = 16
T_LIST_QUICKLIST_2 = 17
T_STREAM_LISTPACKS_2 = 18

def rbyte(f):
    b = f.read(1)
    if not b: raise EOFError("EOF")
    return b[0]

def rbytes(f, n):
    d = f.read(n)
    if len(d) != n: raise EOFError(f"want {n} got {len(d)}")
    return d

def rlen(f):
    first = rbyte(f)
    enc = (first & 0xC0) >> 6
    if enc == 0:  return (first & 0x3F, False)
    if enc == 1:  return (((first & 0x3F) << 8) | rbyte(f), False)
    if enc == 3:
        b = rbytes(f, 8)
        v = 0
        for i in range(8): v |= b[i] << (i*8)
        return (v, False)
    # enc == 2 : special encoding in low 6 bits
    it = first & 0x3F
    if it == 0: return (struct.unpack('<b', rbytes(f,1))[0], True)
    if it == 1: return (struct.unpack('<H', rbytes(f,2))[0], True)
    if it == 2: return (struct.unpack('<I', rbytes(f,4))[0], True)
    raise ValueError(f"Bad special enc: {it}")

def rlen_val(f):
    return rlen(f)[0]

def rstr(f):
    v, enc = rlen(f)
    if enc: return str(v)
    return rbytes(f, v).decode('utf-8', errors='replace')

def rdouble(f):
    return struct.unpack('<d', rbytes(f,8))[0]

def skip(f, t):
    if t == T_STRING: rstr(f)
    elif t in (T_LIST, T_SET):
        c = rlen_val(f)
        for _ in range(c): rstr(f)
    elif t in (T_ZSET, T_ZSET_2):
        c = rlen_val(f)
        for _ in range(c): rstr(f); rdouble(f)
    elif t == T_HASH:
        c = rlen_val(f)
        for _ in range(c): rstr(f); rstr(f)
    elif t in (T_LIST_ZIPLIST, T_SET_INTSET, T_ZSET_ZIPLIST,
               T_HASH_ZIPLIST, T_HASH_LISTPACK, T_ZSET_LISTPACK,
               T_STREAM_LISTPACKS, T_STREAM_LISTPACKS_2):
        rstr(f)  # stored as a string blob
    elif t in (T_LIST_QUICKLIST,):
        c = rlen_val(f)
        for _ in range(c): rstr(f)
    elif t == T_LIST_QUICKLIST_2:
        c = rlen_val(f)
        for _ in range(c):
            container = rbyte(f)
            if container == 2:
                cl = rlen_val(f); ul = rlen_val(f)
                rbytes(f, cl)
            else: rstr(f)
    else:
        raise ValueError(f"Cannot skip type {t}")

def parse_kv(f, t):
    key = rstr(f)
    if t == T_STRING:
        return (key, rstr(f), 'string')
    if t == T_HASH:
        c = rlen_val(f); val = {}
        for _ in range(c):
            field = rstr(f); val[field] = rstr(f)
        return (key, val, 'hash')
    if t == T_SET:
        c = rlen_val(f); val = []
        for _ in range(c): val.append(rstr(f))
        return (key, val, 'set')
    if t in (T_ZSET, T_ZSET_2):
        c = rlen_val(f); val = []
        for _ in range(c):
            val.append({'member': rstr(f), 'score': rdouble(f)})
        return (key, val, 'zset')
    if t == T_LIST:
        c = rlen_val(f); val = []
        for _ in range(c): val.append(rstr(f))
        return (key, val, 'list')
    # For encoded types, just capture size
    if t in (T_LIST_ZIPLIST, T_SET_INTSET, T_ZSET_ZIPLIST,
               T_HASH_ZIPLIST, T_HASH_LISTPACK, T_ZSET_LISTPACK):
        raw = rstr(f)
        return (key, f'<{len(raw)} bytes>', 'unknown')
    if t in (T_LIST_QUICKLIST, T_LIST_QUICKLIST_2):
        c = rlen_val(f)
        parts = []
        for _ in range(c):
            container = rbyte(f)
            if container == 2:
                cl = rlen_val(f); ul = rlen_val(f)
                rbytes(f, cl)
                parts.append('<compressed>')
            else:
                rstr(f)
                parts.append('<ziplist>')
        return (key, parts, 'list')
    if t in (T_STREAM_LISTPACKS, T_STREAM_LISTPACKS_2):
        c = rlen_val(f)
        for _ in range(c): rstr(f)
        return (key, '<stream>', 'stream')
    skip(f, t)
    return (key, f'<type-{t}>', 'unknown')

def parse(rdb_path, targets):
    all_keys = {}
    results = {}
    cur_db = 0
    with open(rdb_path, 'rb') as f:
        magic = f.read(5)
        if magic != b'REDIS': raise ValueError(f"Bad magic {magic}")
        ver = int(f.read(4).decode('ascii'))
        print(f"RDB version: {ver}", file=sys.stderr)

        while True:
            try: opcode = rbyte(f)
            except EOFError: break

            if opcode == RDB_EOF:
                print("EOF.", file=sys.stderr); break

            if opcode == RDB_AUX:
                # AUX field: key + value (both strings)
                rstr(f); rstr(f)
                continue

            if opcode == RDB_SELECTDB:
                cur_db = rlen_val(f); continue

            if opcode == RDB_RESIZEDB:
                rlen(f); rlen(f); continue

            expire = None
            if opcode == RDB_EXPIRETIME_MS:
                expire = rbytes(f, 8)
                opcode = rbyte(f)
            elif opcode == RDB_EXPIRETIME:
                expire = rbytes(f, 4)
                opcode = rbyte(f)

            try:
                key, value, vtype = parse_kv(f, opcode)
                full = f"db{cur_db}:{key}"
                entry = {'key': key, 'db': cur_db, 'type': vtype, 'value': value}
                all_keys[full] = entry
                if any(t in key for t in targets):
                    results[full] = entry
            except Exception as e:
                print(f"Error at {f.tell()}: {e}", file=sys.stderr)
                break

    return all_keys, results

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print(f"Usage: python {sys.argv[0]} <rdb> [user1,user2,...]")
        sys.exit(1)
    rdb = sys.argv[1]
    targets = sys.argv[2].split(',') if len(sys.argv) > 2 else ['7788','5566']
    print(f"Parsing: {rdb}", file=sys.stderr)
    print(f"Targets: {targets}", file=sys.stderr)
    all_keys, results = parse(rdb, targets + ['admin','config','users'])
    print(f"\nTotal keys: {len(all_keys)}", file=sys.stderr)
    print(f"Matches: {len(results)}\n", file=sys.stderr)
    for k, e in results.items():
        print(f"[{e['type']}] {e['key']}")
        s = json.dumps(e['value'], ensure_ascii=False)
        print(f"  = {s[:400]}")
        print()
    if not results:
        print("No matches. Key pattern samples:", file=sys.stderr)
        seen = set()
        for k, e in list(all_keys.items())[:60]:
            p = e['key'].split(':')[0]
            if p not in seen:
                seen.add(p)
                print(f"  [{e['type']}] {e['key'][:70]}", file=sys.stderr)
